### PR TITLE
drivers: uart: Allow to pass arbitrary user data to irq callback

### PR DESCRIPTION
**Pull request by pfalcon**
Zephyr UART drivers offer very low-level functionality. Oftentimes,
it would be useful to provide higher-level wrappers around UART
device which would offer additional functionality. However, UART
driver irq callback routine receives just a pointer to (low-level)
UART device, and it's not possible to get to a wrapper structure
(without introducing expensive external mapping structures). This
is an indirect reason why the current UARt wrappers - uart_pipe,
console - are instantiated statically just for one underlying UART
device and cannot be reused for multiple devices.

Solve this by allowing to pass an arbitrary user data to irq
callback, set by new uart_irq_callback_user_data_set() function.
Existing uart_irq_callback_set() keeps setting a callback which
will receive pointer to the device.

While public API maintains compatibility, drivers themselves need
to be updated to support arbitrary user data storage/passing (as
legacy uart_irq_callback_set() functionality is now implemented in
terms of it).

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

### CODE REVIEW

**Comment by pfalcon** [comment]
@mike-scott : This is response to context_from_dev() cuteness from your https://github.com/zephyrproject-rtos/zephyr/pull/6344/commits/779ba53856a359e4621214fe1118140a6f5d97b3#diff-d06de45a821bbe135283a43006e18f59R38 . (But prompted by my attempt to refactor https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/console/getchar.c to be reusable across devices, and seeing why you had that overhead). We should stop working around things, but finally get them right.

**Comment by pfalcon** [comment]
Note: this is WIP as of time of writing, only a couple of drivers is converted, the rest is to do.

**Comment by pfalcon** [comment]
Note: this is WIP as of time of writing, only a couple of drivers is converted, the rest is to do.

**Comment by pfalcon** [comment]
@andrewboie: As this is prerequisite of https://github.com/zephyrproject-rtos/zephyr/pull/8952, please consider acking, then approving.

**Comment by pfalcon** [comment]
@andrewboie: As this is prerequisite of https://github.com/zephyrproject-rtos/zephyr/pull/8952, please consider acking, then approving.

**Comment by mike-scott** [comment]
@pfalcon I think the idea behind this PR looks good.  I'll look at the code when I get a bit of time.  Thanks.

**Comment by vaussard** [comment]
Hi @pfalcon, in the callback, why not passing both the pointers to the `struct device *` and the user data? Something similar is done in the counter API (and probably others).

**Comment by pfalcon** [comment]
All drivers are now converted, CI passes, this PR is ready for review.

File: include/uart.h
Comment (finikorg) [inline include/uart.h]: This data type sounds like a user data of a callback, very confusing.

Comment (pfalcon) [reply include/uart.h]: > This data type sounds like a user data of a callback, very confusing.

Let's fix it, choose:

1. uart_irq_callback_with_user_data_t
2. uart_irq_callback_w_user_data_t
3. uart_irq_callback_w_ud_t
4. uart_irq_cb_w_ud_t
5. uart_irq_user_data_callback_t
...
20. Your choice.

Comment (mike-scott) [reply include/uart.h]: Why change the name of the callback type at all?
uart_irq_callback_t still seems appropriate to me.  It just takes a parameter now.

Comment (mike-scott) [reply include/uart.h]: I realize this patch is leaving the older APIs in place, but does it need to?  Let me take another look.

Comment (mike-scott) [reply include/uart.h]: @pfalcon To clarify: why are we going to through all this trouble to maintain a legacy version of uart_irq_callback_t (and related functions which use it) when there are no in-tree users and we're changing the UART API for irq_callback_set anyway?  Either way, we're breaking users.

The only way I can see where we don't immediately break users would be to keep the irq_callback_set API call the way it is, and add a new irq_callback_user_data_set API which can be used to set the user_data if needed.  In this case, the exposed function uart_irq_callback_user_data_set() would only take struct device *dev and void *user_data params.

**Review by tyrantlizard77 — COMMENTED** [review]
LGTM
May need a waiver to our deprecation policy as mentioned in the other patch, or introduce new API and deprecate the old one

File: include/uart.h
Comment (pfalcon) [reply include/uart.h]: > To clarify: why are we going to through all this trouble to maintain a legacy version of uart_irq_callback_t

So, there're to side of things:

1. Public API of UART subsystem (i.e. API between UART subsys and user apps).
2. Private (hopefully) API between UART subsys and UART drivers.

Public API of UART subsystem is in use in the tree, even if uart_irq_callback_t does not occur much explicitly. Users include: uart_pipe, subsys/console, some tests, maybe some samples, your https://github.com/zephyrproject-rtos/zephyr/pull/6344. Then there can be out of tree users too. So, a care is taken to not break that usage, so existing uart_irq_callback_t is left as is (as well as actuall callback funcs with the same signature, even if not declared as uart_irq_callback_t), and a new setter and callback type is introduced for user_data-taking callback.

It's different story with private'ish API between subsys and UART drivers. Maintaining compatibility there would be rather cumbersome, so there we switch to uart_irq_callback_user_data_t right away (and older uart_irq_callback_t is handled as a particular special case of it using "ducktyping").

**Comment by pfalcon** [comment]
@andrewboie:

> LGTM
> May need a waiver to our deprecation policy as mentioned in the other patch, or introduce new API and deprecate the old one

As explained in https://github.com/zephyrproject-rtos/zephyr/pull/8951#discussion_r204036199, this exactly introduces new API calls for new functionality on the public API side. But indeed, changes driver-level API. I hope that makes sense - driver API is internal API of Zephyr, and folks who don't submit drivers upstream (where they will be refactored on changes, like this patch does) really should think twice.

So, if this looks good, and you don't have specific suggestions for better names of new API call/type, or something, can you please approve this?

**Comment by andrewboie** [comment]
>  driver API is internal API of Zephyr

This is news to me, private APIs in Zephyr are indicated as such using a naming convention, by prefixing with "z_" or "_".
@nashif what is your call on this one?

**Comment by nashif** [comment]
> driver API is internal API of Zephyr

@pfalcon what driver APIs are you referring to? include/uart.h or something like drivers/serial/uart_ns16550.h?

**Comment by pfalcon** [comment]
@nashif 

>  what driver APIs are you referring to?

I'm specifically referring to `struct uart_driver_api` https://github.com/zephyrproject-rtos/zephyr/blob/master/include/uart.h#L110 , which of course isn't what applications should use, and thus can't have the same status as e.g. uart_fifo_fill() https://github.com/zephyrproject-rtos/zephyr/blob/master/include/uart.h#L269

**Comment by pfalcon** [comment]
> Hi @pfalcon, in the callback, why not passing both the pointers to the struct device * and the user data? Something similar is done in the counter API (and probably others).

Would like to get back to this too. So, taken as just a simple question, likewise a simple answer (why I did it like that) is given in https://github.com/zephyrproject-rtos/zephyr/pull/8951#issuecomment-405519347.

However, let's consider that as a would-be request (I'd like more explicit wording for that to be treated as a request, and preferrably from more than one party, also preferrably with experience in designing such callback signatures in other parts of Zephyr). My analysis however would be:

> Something similar is done in the counter API

Indeed, it had `typedef void (*counter_callback_t)(struct device *dev, void *user_data);`: https://github.com/zephyrproject-rtos/zephyr/blob/master/include/counter.h#L30

However: https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/counter/timer_tmr_cmsdk_apb.c#L144 . D'oh:

~~~
		/* user_data paramenter is not used by this driver */
		(*user_cb)(dev, NULL);
~~~

> (and probably others).

Searching by just "user_data", I don't see APIs for other hardware blocks to use user_data. And based on just info above, I lean towards saying that UART subsys *randomly* doesn't pass user_data at all, while Counter subsys likewise *randomly* passes both dev and user_data. Maybe the fix for UART subsys to pass both dev and user_data, but who knows, maybe it's Counter subsys should be "fixed" to pass only user_data.

What's interesting is that we have concurrent PR to do about the same for DMA subsys: https://github.com/zephyrproject-rtos/zephyr/pull/9018. It has

~~~
-	void (*dma_blkcallback)(struct device *dev, u32_t channel,
+	void (*dma_blkcallback)(void *arg, u32_t channel,
 		int error_code);
~~~

I.e. does just the same what this PR does, in the assumption that in any realistic case, a structure pointed by user_data would contain `struct device *dev` anyway (because we want to use `dev` not just from IRQ callback, but surely from other API calls too).


So, all in all, we have a situation of "original designs", where different persons design similar aspects of similar APIs in different ways. Ideally, there would project maintainer(s) to oversee such matters, and I'm open to suggestion from them, but they should apply them also to https://github.com/zephyrproject-rtos/zephyr/pull/9018/ , to https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/counter/timer_tmr_cmsdk_apb.c#L144 , etc.

**Review by tyrantlizard77 — APPROVED** [review]
I don't have any objections to this as long as @nashif is OK with API change.

**Comment by vaussard** [comment]
>  Would like to get back to this too. So, taken as just a simple question, likewise a simple answer (why I did it like that) is given in #8951 (comment).

Thank you for your detailed answer. That was just a minor question and not a request for changing. My goal was mainly to point out the heterogeneity in the existing APIs for similar problems. It would be good to unify that at some point.

I also thought that it would be easier to keep the pointer to `struct device` because most users are only interested by it. But in that case the user data can be set to point to the `struct device`, so this is not an issue.

**Review by erwango — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
Nice addition, thanks!

**Pull request merged**
